### PR TITLE
txthrottler: verify config at vttablet startup, consolidate funcs

### PR DESCRIPTION
**Pull request by timvaillancourt**
## Description

This PR validates the required params of `txthrottler` earlier in `go/vt/vttablet/tabletserver/tabletenv` to resolve the throttler being ~~silently~~ quietly disabled _(just 1 warning log-line)_ when the expected config is not provided

The current behaviour is not ideal because it is easy to miss that the throttler was disabled and it is inconsistent with how most other `vttablet` features are initialized

Details:
1. Move `txthrottler` config validation to `go/vt/vttablet/tabletserver/tabletenv/config.go`:
    - Added `TxThrottlerConfigFlag` to streamline validation of `--tx-throttler-config` flag
        - I wanted to put this under `go/vt/vttablet/tabletserver/txthrottler` but that causes an import-loop
    - Used `vterrors.New(...)` or `vterrors.Errorf(...)` for errors
    - Moved validation of `TxThrottlerDefaultPriority`/`--tx-throttler-default-priority` to `verifyTxThrottlerConfig()` func
    - Updated unit tests
2. Consolidated `NewTxThrottler()`, `tryCreateTxThrottler()` and `newTxThrottler` into a single func: `NewTxThrottler`
    - The removal of config-validation made the `error` return needless, making the `tryCreateTxThrottler()` + `newTxThrottler` logic needless
    - Updated unit tests

cc @ejortegau / @shlomi-noach 🙏 

## Related Issue(s)

This resolves https://github.com/vitessio/vitess/issues/12410

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/flags/endtoend/vttablet.txt [hunk 1]
```diff
@@ -345,11 +345,10 @@ Usage of vttablet:
...
-      --tx_throttler_config string                                       The configuration of the transaction throttler as a text-formatted throttlerdata.Configuration protocol buffer message. (default "target_replication_lag_sec: 2\nmax_replication_lag_sec: 10\ninitial_rate: 100\nmax_increase: 1\nemergency_decrease: 0.5\nmin_duration_between_increases_sec: 40\nmax_duration_between_increases_sec: 62\nmin_duration_between_decreases_sec: 20\nspread_backlog_across_sec: 20\nage_bad_rate_after_sec: 180\nbad_rate_increase: 0.1\nmax_rate_approach_threshold: 0.9\n")
+      --tx_throttler_config string                                       The configuration of the transaction throttler as a text-formatted throttlerdata.Configuration protocol buffer message. (default "target_replication_lag_sec:2 max_replication_lag_sec:10 initial_rate:100 max_increase:1 emergency_decrease:0.5 min_duration_between_increases_sec:40 max_duration_between_increases_sec:62 min_duration_between_decreases_sec:20 spread_backlog_across_sec:20 age_bad_rate_after_sec:180 bad_rate_increase:0.1 max_rate_approach_threshold:0.9")
```
Comment (timvaillancourt) [inline go/flags/endtoend/vttablet.txt:351]: Looks like a custom flag removes the `--tx-throttler-config` "synonym". Ideas appreciated if this is a concern

File: go/vt/vttablet/tabletserver/tabletenv/config.go [hunk 2]
```diff
@@ -657,9 +677,6 @@ func (c *TabletConfig) Verify() error {
 	if v := c.HotRowProtection.MaxConcurrency; v <= 0 {
 		return fmt.Errorf("--hot_row_protection_concurrent_transactions must be > 0 (specified value: %v)", v)
 	}
-	if v := c.TxThrottlerDefaultPriority; v > sqlparser.MaxPriorityValue || v < 0 {
```
Comment (timvaillancourt) [inline go/vt/vttablet/tabletserver/tabletenv/config.go:660]: Moving this to `verifyTxThrottlerConfig()`

File: go/vt/vttablet/tabletserver/tabletenv/config_test.go [hunk 3]
```diff
@@ -332,31 +334,101 @@ func TestFlags(t *testing.T) {
...
+	defaultMaxReplicationLagModuleConfig := throttler.DefaultMaxReplicationLagModuleConfig().Configuration
+
 	{
-		// default config (replica)
+		// default (disabled)
 		assert.Nil(t, currentConfig.verifyTxThrottlerConfig())
 	}
 	{
-		// replica + rdonly (allowed)
+		// enabled without throttler config
```
Comment (timvaillancourt) [inline go/vt/vttablet/tabletserver/tabletenv/config_test.go:365]: This could be made a `for`-loop-based test, let me know

Comment (timvaillancourt) [reply go/vt/vttablet/tabletserver/tabletenv/config_test.go]: I did this anyways ✅

File: go/flags/endtoend/vttablet.txt (hunk 1 above)
Comment (timvaillancourt) [reply go/flags/endtoend/vttablet.txt:351]: cc @ajm188 for thoughts on the above ^^

File: go/vt/vttablet/tabletserver/tabletenv/config.go [hunk 4]
```diff
@@ -158,7 +178,7 @@ func registerTabletEnvFlags(fs *pflag.FlagSet) {
 	SecondsVar(fs, &currentConfig.TwoPCAbandonAge, "twopc_abandon_age", defaultConfig.TwoPCAbandonAge, "time in seconds. Any unresolved transaction older than this time will be sent to the coordinator to be resolved.")
 	// Tx throttler config
 	flagutil.DualFormatBoolVar(fs, &currentConfig.EnableTxThrottler, "enable_tx_throttler", defaultConfig.EnableTxThrottler, "If true replication-lag-based throttling on transactions will be enabled.")
-	flagutil.DualFormatStringVar(fs, &currentConfig.TxThrottlerConfig, "tx_throttler_config", defaultConfig.TxThrottlerConfig, "The configuration of the transaction throttler as a text-formatted throttlerdata.Configuration protocol buffer message.")
+	fs.Var(currentConfig.TxThrottlerConfig, "tx_throttler_config", "The configuration of the transaction throttler as a text-formatted throttlerdata.Configuration protocol buffer message.")
```
Comment (harshit-gangal) [inline go/vt/vttablet/tabletserver/tabletenv/config.go:181]: change the method from `DualFormatStringVar` to `Var` removed the flag with `-` . I think we should not do that unless there is a strong reason for it.

**Review by harshit-gangal — APPROVED** [review]

File: go/vt/vttablet/tabletserver/tabletenv/config.go (hunk 4 above)
Comment (timvaillancourt) [reply go/vt/vttablet/tabletserver/tabletenv/config.go]: @harshit-gangal good point, I've added a `DualFormatVar` to `go/flagutil` and moved to this 👍

File: go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go [hunk 5]
```diff
@@ -186,64 +184,36 @@ type txThrottlerState struct {
...
-	if !env.Config().EnableTxThrottler {
-		return newTxThrottler(env, topoServer, &txThrottlerConfig{enabled: false})
-	}
+	throttlerConfig := &txThrottlerConfig{enabled: false}
+
+	if env.Config().EnableTxThrottler {
+		// Clone tsv.TxThrottlerHealthCheckCells so that we don't assume tsv.TxThrottlerHealthCheckCells
+		// is immutable.
+		healthCheckCells := make([]string, len(env.Config().TxThrottlerHealthCheckCells))
+		copy(healthCheckCells, env.Config().TxThrottlerHealthCheckCells)
```
Comment (shlomi-noach) [inline go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go:193]: My personal preference is to use a simpler one-liner:
```suggestion
      healthCheckCells := append(env.Config().TxThrottlerHealthCheckCells)
```

**Review by shlomi-noach — APPROVED** [review]

File: go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go (hunk 5 above)
Comment (timvaillancourt) [reply go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go:193]: @shlomi-noach SGTM thanks 👍

Comment (timvaillancourt) [reply go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go]: @shlomi-noach looks like this fails the Static Code analysis CI with:

```bash
Error: go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go:192:23: SA4021: x = append(y) is equivalent to x = y (staticcheck)
		healthCheckCells := append(env.Config().TxThrottlerHealthCheckCells)
```

So I'll go with what the linter is suggesting. I think defining a new var will create a copy too

Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go]: Huh, OK TIL.

> The underlying array is reused if there is enough capacity

I didn't think this through. And of course it makes perfect sense.

**Pull request merged**